mermithid/processors/misc/ComplexLineShape.py

```diff
@@ -0,0 +1,598 @@
+        full_spectrum = 0
+        for i in range(start_number_of_i, end_number_of_i, 1):
+            full_spectrum += self.spectrum1(i, E)
+        return full_spectrum
+
+    # shake spectrum by adding up shake spectrum for all the states up to i=24
+    def shake_spectrum(self):
+        x_array = self.input_std_eV_array
+        x_array = flip_array(x_array)
+        shake_spectrum = self.full_shake_spectrum(x_array, 0, 24)
+        return shake_spectrum
+    ###############################################################################
+
+class ComplexLineShape(BaseProcessor):
+
+    def InternalConfigure(self, params):
+        '''
+        Configure
+        '''
+        # Read other parameters
+        self.bins_choice = reader.read_param(params, 'bins_choice', [])
+        self.gases = reader.read_param(params, 'gases', ["H2","Kr"])
+        self.max_scatters = reader.read_param(params, 'max_scatters', 20)
+        self.max_comprehensive_scatters = reader.read_param(params, 'max_comprehensive_scatters', 4)
+        # This is an important parameter which determines how finely resolved
+        # the scatter calculations are. 10000 seems to produce a stable fit, with minimal slowdown
+        self.num_points_in_std_array = reader.read_param(params, 'num_points_in_std_array', 10000)
+        self.RF_ROI_MIN = reader.read_param(params, 'RF_ROI_MIN', 25850000000.0)
+        self.B_field = reader.read_param(params, 'B_field', 0.957810722501)
+        self.shake_spectrum_parameters_json_path = reader.read_param(params, 'shake_spectrum_parameters_json_path', 'shake_spectrum_parameters.json')
+        self.path_to_osc_strengths_files = reader.read_param(params, 'path_to_osc_strengths_files', '/host/')
+
+    def InternalRun(self):
+
+        # Read shake parameters from JSON file
+        self.shakeSpectrumClassInstance = ShakeSpectrumClass(self.shake_spectrum_parameters_json_path, self.std_eV_array())
+
+        # number_of_events = len(self.data['StartFrequency'])
+        # self.results = number_of_events
+
+        a = self.data['StartFrequency']
+
+        # fit with shake spectrum
+        data_hist_freq, freq_bins= np.histogram(a,bins=self.bins_choice)
+        histogram = data_hist_freq
+        bins = freq_bins
+        guess = np.where(np.array(histogram) == np.max(histogram))[0][0]
+        print(guess)
+        kr17kev_in_hz = guess*(bins[1]-bins[0])+bins[0]
+        #self.B_field = B(17.8, kr17kev_in_hz + 0)
+        self.results = self.fit_data(freq_bins, data_hist_freq, self.shakeSpectrumClassInstance)
+
+        return True
+
+
+    # Establishes a standard energy loss array (SELA) from -1000 eV to 1000 eV
+    # with number of points equal to self.num_points_in_std_array. All convolutions
+    # will be carried out on this particular discretization
+    def std_eV_array(self):
+        emin = -1000
+        emax = 1000
+        array = np.linspace(emin,emax,self.num_points_in_std_array)
```

Comment:
it is probably wise to make sure that array contains 0. (within numerical precision). If it doesn't and we convolve this lineshape (with delta function as base function instead of the shake-up/shake-off spectrum) with the tritium spectrum, we will shift the spectrum by the distance between 0 and the entry in array closest to zero.